client/ayon_flame/plugins/publish/collect_shots.py

```diff
@@ -0,0 +1,339 @@
+from ayon_flame.otio import flame_export
+
+from ayon_core.pipeline.editorial import (
+    get_media_range_with_retimes
+)
+
+
+# constatns
+NUM_PATERN = re.compile(r"([0-9\.]+)")
+TXT_PATERN = re.compile(r"([a-zA-Z]+)")
+
+
+class CollectShot(pyblish.api.InstancePlugin):
+    """Collect new shots."""
+
+    order = order = pyblish.api.CollectorOrder - 0.095
+    label = "Collect Shots"
+    hosts = ["flame"]
+    families = ["shot"]
+
+    SHARED_KEYS = (
+        "folderPath",
+        "fps",
+        "handleStart",
+        "handleEnd",
+        "item",
+        "otioClip",
+        "resolutionWidth",
+        "resolutionHeight",
+        "retimedHandles",
+        "retimedFramerange",
+        "path",
+        "pixelAspect",
+        "sourceFirstFrame",
+        "versionData",
+        "workfileFrameStart",
+        "xml_overrides",
+    )
+
+    # TODO: add to own plugin for Flame
+    # TODO: toggle for marking task which should be used for product parent
+    add_tasks = []    
+
+    @classmethod
+    def _inject_editorial_shared_data(cls, instance):
+        """
+        Args:
+            instance (obj): The publishing instance.
+        """
+        context = instance.context
+        instance_id = instance.data["instance_id"]
+
+        # Inject folderPath and other creator_attributes to ensure
+        # new shots/hierarchy are properly handled.
+        creator_attributes = instance.data['creator_attributes']
+        instance.data.update(creator_attributes)
+
+        # Inject/Distribute instance shot data as editorialSharedData
+        # to make it available for clip/plate/audio products
+        # in sub-collectors.
+        if not context.data.get("editorialSharedData"):
+            context.data["editorialSharedData"] = {}
+
+        context.data["editorialSharedData"][instance_id] = {
+            key: value for key, value in instance.data.items()
+            if key in cls.SHARED_KEYS
+        }
+
+    def process(self, instance):
+        """
+        Args:
+            instance (pyblish.Instance): The shot instance to update.
+        """
+        instance.data["integrate"] = False  # no representation for shot
+
+        # Adjust instance data from parent otio timeline.
+        otio_timeline = instance.context.data["otioTimeline"]
+        otio_clip, marker = self.get_marker_from_clip_index(
+            otio_timeline, instance.data["clip_index"]
+        )
+        if not otio_clip:
+            raise RuntimeError("Could not retrieve otioClip for shot %r", instance)
+
+        # Compute fps from creator attribute.
+        creator_attrs = instance.data['creator_attributes']
+        if creator_attrs["fps"] == "from_selection":
+            creator_attrs["fps"] = instance.context.data["fps"]
+
+        # Retrieve AyonData marker for associated clip.
+        instance.data["otioClip"] = otio_clip
+
+        # Compute additional data
+        for segment_item in instance.context.data["flameSelectedSegments"]:
```

Comment:
This looks dangerous:
1.) If `instance.context.data["flameSelectedSegments"]` is empty then `segment_item` is not set.
2.) Last segment item is used if none of them matches `clip_index` from instance.

Reply:
Yes you're right, I have added a raise condition as a `for-else` this way the plugin does not go further if source segment cannot be explicitely retrieved from the clip index.